core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java

```diff
@@ -0,0 +1,99 @@
+ */
+package com.linecorp.armeria.client.endpoint.healthcheck;
+
+import java.net.StandardProtocolFamily;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.ClientOptionsBuilder;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.HttpClient;
+import com.linecorp.armeria.client.HttpClientBuilder;
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.AsyncCloseable;
+
+final class HttpHealthChecker implements AsyncCloseable {
+
+    private final HealthCheckerContext ctx;
+    private final HttpClient httpClient;
+    private final String path;
+    @Nullable
+    private CompletableFuture<?> lastCheckFuture;
+
+    HttpHealthChecker(HealthCheckerContext ctx, String path) {
+
+        final Endpoint endpoint = ctx.endpoint();
+        final SessionProtocol protocol = ctx.protocol();
+        final String scheme = protocol.uriText();
+        final String ipAddr = endpoint.ipAddr();
+        final HttpClientBuilder builder;
+        if (ipAddr == null) {
+            builder = new HttpClientBuilder(scheme + "://" + endpoint.authority());
+        } else {
+            final int port = ctx.port() > 0 ? ctx.port() : endpoint.port(protocol.defaultPort());
+            if (endpoint.ipFamily() == StandardProtocolFamily.INET) {
+                builder = new HttpClientBuilder(scheme + "://" + ipAddr + ':' + port);
+            } else {
+                builder = new HttpClientBuilder(scheme + "://[" + ipAddr + "]:" + port);
+            }
+            builder.setHttpHeader(HttpHeaderNames.AUTHORITY, endpoint.authority());
+        }
+
+        this.ctx = ctx;
+        httpClient = builder.factory(ctx.clientFactory())
+                            .options(ctx.clientConfigurator().apply(new ClientOptionsBuilder()).build())
+                            .build();
+        this.path = path;
+    }
+
+    void start() {
+        check();
+    }
+
+    private void check() {
+        final CompletableFuture<AggregatedHttpResponse> f;
+        synchronized (this) {
+            f = httpClient.head(path).aggregate();
+            lastCheckFuture = f;
+        }
+
+        f.handle((res, cause) -> {
+            if (res != null && res.status().equals(HttpStatus.OK)) {
+                ctx.updateHealth(1);
+            } else {
+                ctx.updateHealth(0);
+            }
+
+            ctx.executor().schedule(this::check, ctx.nextDelayMillis(), TimeUnit.MILLISECONDS);
+            return null;
+        });
+    }
+
+    @Override
+    public synchronized CompletableFuture<?> closeAsync() {
+        if (lastCheckFuture != null) {
+            return lastCheckFuture.handle((unused1, unused2) -> null);
```

Comment:
I think it's possible for this to complete before an in-progress check is completed since `f.handle` above is not assigned to `lastCheckFuture`. Is that ok?

Reply:
A little related, wonder if `f.handle` above should check whether the checker is closed or not. I'm not sure I see a mechanism that prevents `ctx.executor().schedule(this::check, ctx.nextDelayMillis(), TimeUnit.MILLISECONDS);` from continuing to happen forever even after close is called.

Reply:
> I think it's possible for this to complete before an in-progress check is completed since f.handle above is not assigned to lastCheckFuture. Is that ok?

Nice catch! :sweat_smile: 

> A little related, wonder if f.handle above should check whether the checker is closed or not.

It is now impossible to schedule a task on an executor of a destroyed context, so it will be OK, but I think we need to catch `RejectedExecutionException` so that the log is not polluted. Thanks for the hints!

Reply:
Should be fixed now. Had to merge `DefaultHealthCheckerContext` and `CancellableExecutor` to simplify destruction process.